### PR TITLE
Add painless string split function (splitOnToken)

### LINKED ISSUE #20952: Painless - Request for native String split function
**Describe the feature**:
String.split is not exposed for performance.  Can achieve using regex but this requires the user to enable regexes for what is a rather common use case.

Can we have a native split that only supports a char sequence and not a regex?

### PR DESCRIPTION
**Pull request by cam72cam**
Fixes #20952

Example Usage:
```json
{
 "pipeline": {
   "processors": [
     {
       "script": {
         "source": "ctx.ip = ctx.ip.splitOnToken(',')[0]"
       }
     },
     {
       "geoip": {
         "field": "ip",
         "target_field": "geoip"
       }
     }
   ]
 },
 "docs": [
   {
     "_source": {
       "message": "hello world",
       "ip": "130.167.201.50"
     }
   },
   {
     "_source": {
       "message": "hello world",
       "ip": "130.167.201.50, 130.167.201.51"
     }
   }
 ]
}
```

### CODE REVIEW

**Comment by jdconrad** [comment]
@cam72cam Thank you for submitting this.  However, in Painless we have taken great care to avoid compiling regexes in all but static cases since compilation can have very poor performance.  If you're still interested, perhaps you could re-implement this set of methods using StringTokenizer or something similar?  The method names and whitelisting both look good.

**Comment by cam72cam** [comment]
Ah, I understand.  Even though I am escaping a regex, java is still using a dynamically compiled regex behind the scenes.  I'll take a stab at refactoring this next week.

**Comment by jdconrad** [comment]
@cam72cam Just wanted to loop back with you and see if you still had interest in continuing this PR?  Thanks!

**Comment by cam72cam** [comment]
I've finally got some time to get back on this.  I'll try to get the branch updated before the weekend

**Comment by cam72cam** [comment]
@jdconrad I've pushed my own custom implementation of this functionality.  Based on my testing it has similar performance for single char tokens and better performance on longer tokens.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java [hunk 1]
```diff
@@ -503,4 +503,53 @@ public static String encodeBase64(String receiver) {
...
+            result.add(receiver.substring(pos, idx));
+
+            // Move our search position to the next possible location
+            pos = idx + token.length();
+        }
+        // Add the remaining string to the result list
+        result.add(receiver.substring(pos));
+
+        // O(N) or faster depending on implementation
+        return result.toArray(new String[0]);
```
Comment (jdconrad) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java]: If you make the ```new String[0]``` into ```new String[results.size()]``` the array will not have to be allocated twice.

**Review by jdconrad — CHANGES_REQUESTED** [review]
@cam72cam Thanks for updating this.  It's starting to look good.  Would you please add unit tests for these two methods?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java [hunk 2]
```diff
@@ -503,4 +503,53 @@ public static String encodeBase64(String receiver) {
...
+        ArrayList<String> result = new ArrayList<String>();
+
+        // Keep track of where we are in the string
+        // indexOf(tok, startPos) is faster than creating a new search context ever loop with substring(start, end)
+        int pos = 0;
+
+        // Loop until we hit the limit or forever if we are passed in less than one (signifying no limit)
+        // If Integer.MIN_VALUE is passed in, it will still continue to loop down to 1 from MAX_VALUE
+        // This edge case should be fine as we are limited by receiver length (Integer.MAX_VALUE) even if we split at every char
+        for(;limit != 1; limit--) {
```
[the remark this answers is not in the document]
Comment (cam72cam) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java]: That is the desired functionality and matches String.split in terms of accepted length.  less than 1 should be treated as split as many times as you can.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java [hunk 3]
```diff
@@ -503,4 +503,53 @@ public static String encodeBase64(String receiver) {
...
+
+    /**
+     * Split 'receiver' by 'token' as many times as possible..
+     */
+    public static String[] splitOnToken(String receiver, String token) {
+        return splitOnToken(receiver, token, -1);
+    }
+
+    /**
+     * Split 'receiver' by 'token' up to 'limit' times.  Any limit less than 1 is ignored.
```
[the remark this answers is not in the document]
Comment (cam72cam) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java]: See above comment.  I'd like to keep this as compatible with String.split as possible

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java (hunk 1 above)
Comment (cam72cam) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java]: From what I have been reading, it's more efficient to use a empty array here:
https://stackoverflow.com/questions/5374311/convert-arrayliststring-to-string-array
https://shipilev.net/blog/2016/arrays-wisdom-ancients/

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java]: Yes, of course, brain glitch on my part.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java]: Same as above.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java (hunk 1 above)
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java]: Interesting articles.  Thank you!

**Comment by jdconrad** [comment]
@cam72cam Thanks for the responses.  Just need some unit tests, and then I will run our test suite against this change.

**Comment by cam72cam** [comment]
Ok, I'll try to get some unit tests in next week

**Comment by jdconrad** [comment]
@cam72cam Are you still interested in finishing this up with some unit tests?  Thanks!

**Comment by cam72cam** [comment]
Yes, sorry.  Work is quite hectic and I have not had a chance to sit down for a hour or two at work and write them up.

**Comment by cam72cam** [comment]
@jdconrad I have added some tests and rebased on top of master

**Comment by jdconrad** [comment]
@cam72cam The added tests look good.  Thank you for finishing this up.  However, I'm not sure the rebase went correctly as there are a large number of files that shouldn't have changes.  Would you please try syncing this again with master?

**Comment by cam72cam** [comment]
@jdconrad Looks like someone force pushed to origin/master a few days ago.  I rebased and cleaned up my branch.

**Review by jdconrad — APPROVED** [review]

**Comment by jdconrad** [comment]
@cam72cam This is great, thank you!  Running tests now.

**Comment by jdconrad** [comment]
@elasticmachine test this please

**Comment by cam72cam** [comment]
hmm: 
```
15:29:40 Forbidden field access: java.lang.System#out [prints to System.out; should only be used for debugging, not in production code]
15:29:40   in org.elasticsearch.painless.AugmentationTests (AugmentationTests.java:235)
```

I can comment that out in the tests

**Comment by jdconrad** [comment]
@cam72cam I missed the println when reviewing the tests here.  Please go ahead and remove that line completely, and then I will retest.

**Comment by cam72cam** [comment]
@jdconrad pushed

**Comment by jdconrad** [comment]
@elasticmachine test this please

**Comment by jdconrad** [comment]
@cam72cam Triggered another forbidden API.  I would just use the build in string concat's ```+``` here instead of format.  I don't know what your exact setup is for testing, but if you run the gradle command precommit within the modules/lang-painless directory, it should reveal any other issues related to style.

**Comment by jdconrad** [comment]
@elasticmachine test this please

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/1

**Comment by jdconrad** [comment]
@cam72cam Your change has been merged.  Thank you for your contribution, and thank you for sticking with it to completion!

**Comment by cam72cam** [comment]
Thanks for being patient with my chaotic schedule!

**Pull request merged**